### PR TITLE
Update AddContactPage

### PR DESCRIPTION
**Pull request by golf1052**
It looks kind of jank but it's way better. It first pulls the user's contacts then checks Signal to see which contacts are using Signal. The Signal contacts are at the top of the list. The user also has the option to add a manual contact.

Another way we could handle adding manual contacts would be to just have the search box and if the user enters a valid number they could add that as a contact and set the contact name later.

cc @Trolldemorted 

On desktop
Imgur

On mobile
Imgur

### CODE REVIEW

**Comment by Trolldemorted** [comment]
Will test it in the evening at home, my pc at work is currently on virtualized_unreadmarkers. Why does gh say "This branch cannot be rebased due to conflicts" though? This branch appears like it is just 5 commits ahead of master.

Some thoughts:
- is querying the network for every contact a good idea? A friend of mine is using a windows phone with more than 1000 corporate contacts on it, I am note sure what `accountManager.getContacts` will do in such a case
- since this page now uses the network, should we add a cancellation token to it (and `getContacts`) and cancel it when navigating away?
- do we want non-signal users to be in the list?

**Comment by golf1052** [comment]
I usually merge changes in from master which is probably why the change can't be rebased onto master. I noticed with my last PR that the repo was set to rebase only. In the future should I only rebase instead of merge? The change can be merged in.

- is querying the network for every contact a good idea? A friend of mine is using a windows phone with more than 1000 corporate contacts on it, I am note sure what accountManager.getContacts will do in such a case

Signal-Android does the same thing where it checks the contacts for people on Signal. It probably only checks every once in a while though. We should save the list of known Signal contacts either in the DB or I think there might be an API to create a contact DB with the system.

- since this page now uses the network, should we add a cancellation token to it (and getContacts) and cancel it when navigating away?

Yes

- do we want non-signal users to be in the list?

That's what Signal-Android does, I assume it's because you can use SMS through the app as well. Since we won't have access to the SMS APIs there's probably no point in including non-Signal contacts.

**Comment by Trolldemorted** [comment]
I have seen OWS using merge-by-rebase and I really like it - no merge commits and every merge must be rebased exactly on the current master.

We might want to poll the whole contact list only explicitly by reacting to a user command, but we maybe should read into how Signal-Android handles it first.

**Comment by golf1052** [comment]
Ok I'll try rebasing my changes from now on.

Signal-Android calls `getContacts` on first launch and whenever the user pulls to refresh on the add new contact page.

I've done something similar in my latest commit. I annotate the system contacts when we know they are on Signal so we don't need to ask Signal again about them. We could modify the code further to only call `getContacts` when there are no annotations. If we have annotations then we know we've checked already. The user can then pull to refresh to check again. At some point we should remove the call to `getContacts` on every page load and only check once in a while.

I also had to upgrade the project because my NuGet package references weren't being updated. I think it may have been do to me upgrading to FCU as well :|

File: Signal-Windows/ViewModels/AddContactPageViewModel.cs [hunk 1]
```diff
@@ -12,137 +12,331 @@
...
+        /// Parses and formats a number in E164 format
+        /// </summary>
+        /// <param name="number">The number to parse</param>
+        /// <exception cref="NumberParseException"></exception>
+        /// <returns>A number in E164 format</returns>
+        private string ParsePhoneNumber(string number)
+        {
+            // on phone we should try to get their SIM country code
+            // otherwise we should try to use the user's location?
+            PhoneNumber phoneNumber = phoneNumberUtil.Parse(number, "US");
```
Comment (SERVCUBED) [inline Signal-Windows/ViewModels/AddContactPageViewModel.cs]: This seems to have problems with numbers starting with a '0' (no country code specified) for me. I fixed it in testing with this:

            if (number.StartsWith("0"))
                number = Utils.GetCountryCode() + number.Substring(1, number.Length - 1);

**Pull request merged**